### PR TITLE
chore: chronos in monorepo

### LINKED ISSUE #8335: Move action-chronos into monorepo
Priority: Done by middle of June.

Chatting with @rafaelromcar-parabol we need to handle action-chronos before moving to GCP. We want to move action-chronos into our monorepo so we stop forgetting about it, which has been the source of bugs in the past.

Background:
- Do not want to move to K8s CronJobs. This would tightly couple us to k8s, and we have some clients who only deploy using docker
- The solution we arrived at is treating chronos as a microservice, like we do with our web server & gql executor. 
- by redeploying chronos whenever we redeploy the app we run the risk of a job failing (e.g. a job runs at 5AM & we redeploy at 4:59:59). Essentially this is a MQ problem and we want to achieve "exactly once" calls. If we have multiple instances up, it could get called at least once. If it goes down, it could run 0 times. Our goal is to have the old version running until the new version comes up & then make an immediate switch.

AC
- [ ] chronos is brought into our monorepo as a separate package
- [ ] the prod webpack server config builds chronos & outputs it to /dist/chronos.js
- [ ] running `node dist/chronos.js` will run the chronos scheduler.

### PR DESCRIPTION
**Pull request by mattkrick**
BEFORE MERGING:
- [x] Add env vars to staging & prod

BEFORE DEPLOYING:
- Kill the old chronos service

# Description

fixes #8335

Adds action-chronos to our monorepo

## Demo

https://www.loom.com/share/d98da4b330c24e308ef48b7e8ec44539?sid=14a967c4-93f6-4bce-b391-f87c2c875c1a

## Testing scenarios

See video for tests performed. Just make sure a call from chronos results in a query or mutation getting called

### CODE REVIEW

**Comment by mattkrick** [comment]
cc @rafaelromcar-parabol this adds the chronos microservice. treat it just like the gqlExecutor, just call it with `node dist/chronos`

**Review by igorlesnenko — APPROVED** [review]
Works!

File: packages/chronos/chronos.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+//   Asterisks: all
+//      Ranges: e.g. 1-3,5
+//      Steps: e.g. */2
+
+import {CronJob} from 'cron'
+import getGraphQLExecutor from 'parabol-server/utils/getGraphQLExecutor'
+import publishWebhookGQL from 'parabol-server/utils/publishWebhookGQL'
+
+const chronos = () => {
+  const {CHRONOS_PULSE_EMAIL, CHRONOS_PULSE_CHANNEL, SERVER_ID} = process.env
```
[the remark this answers is not in the document]
Comment (rafaelromcar-parabol) [reply packages/chronos/chronos.ts:23]: Chronos will run only when called. Meaning that we will have multiple copies of the webserver and the gql-executor running, but only one copy of chronos.

Each container in Kuberntes (which runs in a pod) only runs one single process (script).

**Comment by rafaelromcar-parabol** [comment]
That's an amazing work! Very easy and straight forward to make it work.

- **Could you profit from this opportunity to start a small documentation on the different components on the application with a description and the required and optional variables?** I think that would be super useful for us but also for self-hosted clients. That could be added to the `packages/chronos/README.md`. A section like **Environment variables** with a list maybe.
- **Could you add a variable for the time zone?** it would add some flexibility. You could also set it so the default value is whatever you want (I would put UTC as default).
- **Could you add variables to enable and disable each cronjob?** Remeber the incident we had when we were in Lisbon? now that we use Docker images we would need to rebuild a new image if we want to disable one single job. Too much time IMHO.
- **Could you add a variable per cronjob for its cronTime?** that would add sooo much flexibility.

**Comment by mattkrick** [comment]
todo:
- [ ] add env vars used to readme
- [ ] remove timezone in favor of UTC

IN THE FUTURE, IF REQUIRED:
- [ ] we set up 1 ENV VAR per cron job that accepts a cronTime, if no cronTime is there, then the job doesn't run

File: packages/chronos/chronos.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  if (!SERVER_ID) throw new Error('Missing Env Var: SERVER_ID')
+  const canPulse = !!CHRONOS_PULSE_EMAIL && !!CHRONOS_PULSE_CHANNEL
+  const timeZone = 'America/New_York'
+
+  // listen to responses
+  getGraphQLExecutor().subscribe()
+
+  new CronJob({
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/chronos/chronos.ts:31]: yep, those were unused 
- `pingActionTick` was replaced with datadog calls
- we no longer support pausing users, so autopause can safely go away (if i'm wrong please let me know & i can add it back in!)

File: pm2.config.js [hunk 3]
```diff
@@ -1,5 +1,20 @@
 module.exports = {
   apps: [
+    {
+      name: 'Chronos',
+      script: 'dist/chronos.js',
+      instances: 1, // there can only be 1
```
Comment (mattkrick) [inline pm2.config.js:6]: @igorlesnenko today, this is what's making sure we only have 1 chronos running at a time. 
in the future, like @rafaelromcar-parabol says, we'll ditch pm2 for k8s & only deploy a single pod

File: packages/chronos/chronos.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/chronos/chronos.ts:23]: great question! yep, only 1 at a time. tagged you inline https://github.com/ParabolInc/parabol/pull/8429/files#r1244267105

**Comment by rafaelromcar-parabol** [comment]
@mattkrick how can we test that in an on-demand environment? 🤔

**Comment by mattkrick** [comment]
@rafaelromcar-parabol i think we chatted about this sync, but i'll answer here for posterity:
- when action chronos is running it'll console log that it's up & running
- to test any cron job in particular, you can manually copy/paste it into /admin/graphql
- to make sure chronos can talk to redis-- we don't test for that, yet! we can create a ping cronjob, but it will likely fail if it is spun up at the same time as the GQL executor since it'll start up faster. if we need to test for this in the future we can add in some logic like that (e.g. make sure it gets a ping response within 60 seconds of startup)

**Pull request merged**